### PR TITLE
Improve types coverage in Products module

### PR DESCRIPTION
**Pull request by krzysztofwolski**
I want to merge this change because...

- ref #4720
- types matter

### Pull Request Checklist

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [ ] Database queries are optimized and the number of queries is constant.
1. [ ] Database migration files are up to date.
1. [ ] The changes are tested.
1. [ ] GraphQL schema and type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by django-queries** [comment]
Here is the report for 78f8c624d3370c18f18bd9186c040e9ea21fb43f (krzysztofwolski/saleor @ mypy-overload)
Base comparison is aec8d5cc933b1614601115091a69b7eb5af9814d.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  add billing address to checkout            	         34	         34	             20
  add shipping to checkout                   	          7	          7	              0
  checkout payment charge                    	         14	         14	              0
  complete checkout                          	          8	          8	              0
  create checkout                            	         50	         50	             24

# api.benchmark homepage
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  retrieve main menu                         	          5	          5	              0
  retrieve product list                      	          4	          4	              0
  retrieve secondary menu                    	          5	          5	              0
  retrieve shop                              	          2	          2	              0

# api.benchmark product
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product details                            	         15	         15	              3
  retrieve product attributes                	          9	          9	              0

# api.benchmark variant
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product variant bulk create                	         51	         51	              3
  retrieve variant list                      	         15	         15	              6

# api product sorting attributes
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  sort product not having attribute data     	         21	         21	              0
```

</p>
</details>

File: saleor/product/utils/variants_picker.py [hunk 1]
```diff
@@ -10,11 +11,22 @@
 from ...discount import DiscountInfo
 from ...extensions.manager import get_extensions_manager
 from ...seo.schema.product import variant_json_ld
-from ..models import AssignedVariantAttribute, ProductVariant
 from .availability import get_product_availability
 
+if TYPE_CHECKING:
+    # flake8: noqa
```
Comment (NyanKiyoshi) [inline saleor/product/utils/variants_picker.py]: Is it needed?

File: saleor/product/utils/variants_picker.py [hunk 2]
```diff
@@ -37,29 +49,179 @@ def _attributes_to_map(instance: ProductVariant) -> Dict[str, str]:
...
+    slug: str
+    values: List[VariantValues]
+
+    def as_dict(self):
+        return {
+            "pk": self.pk,
+            "name": self.name,
+            "slug": self.slug,
+            "values": [asdict(v) for v in self.values],
+        }
```
Comment (IKarbowiak) [inline saleor/product/utils/variants_picker.py]: Do we need that? It seems that `asdict` from `typing` do exactly the same.

File: saleor/product/utils/variants_picker.py [hunk 3]
```diff
@@ -37,29 +49,179 @@ def _attributes_to_map(instance: ProductVariant) -> Dict[str, str]:
...
+@dataclass
+class MoneyRangeAsDict:
+    minPrice: Optional[MoneyAsDict]
+    maxPrice: Optional[MoneyAsDict]
+
+    def as_dict(self) -> dict:
+        return {
+            "minPrice": asdict(self.minPrice) if self.minPrice else None,
+            "maxPrice": asdict(self.maxPrice) if self.maxPrice else None,
+        }
```
Comment (IKarbowiak) [inline saleor/product/utils/variants_picker.py]: Won't it be easier this way?
```
@dataclass
class MoneyRangeAsDict:
    minPrice: Optional[MoneyAsDict] = None
    maxPrice: Optional[MoneyAsDict] = None
```

File: saleor/product/utils/variants_picker.py [hunk 4]
```diff
@@ -37,29 +49,179 @@ def _attributes_to_map(instance: ProductVariant) -> Dict[str, str]:
...
+        )
+        return {
+            "id": self.id,
+            "availability": self.availability,
+            "price": price,
+            "priceUndiscounted": priceUndiscounted,
+            "attributes": self.attributes,
+            "priceLocalCurrency": priceLocalCurrency,
+            "schemaData": self.schemaData,
+        }
```
Comment (IKarbowiak) [inline saleor/product/utils/variants_picker.py]: As above, I think it's not necessary and could be handled be default values in fields definitions.

File: saleor/product/utils/variants_picker.py [hunk 5]
```diff
@@ -37,29 +49,179 @@ def _attributes_to_map(instance: ProductVariant) -> Dict[str, str]:
...
+            if self.priceRangeLocalCurrency
+            else None
+        )
+        return {
+            "discount": discount,
+            "taxRate": self.taxRate,
+            "priceRange": priceRange,
+            "priceRangeUndiscounted": priceRangeUndiscounted,
+            "priceRangeLocalCurrency": priceRangeLocalCurrency,
+        }
```
Comment (IKarbowiak) [inline saleor/product/utils/variants_picker.py]: As above.

File: saleor/product/utils/variants_picker.py [hunk 6]
```diff
@@ -37,29 +49,179 @@ def _attributes_to_map(instance: ProductVariant) -> Dict[str, str]:
...
+    variantAttributes: List[VariantAttributes] = field(default_factory=list)
+    variants: List[VariantData] = field(default_factory=list)
+
+    def as_dict(self):
+        return {
+            "availability": self.availability.as_dict(),
+            "priceDisplay": asdict(self.priceDisplay),
+            "variantAttributes": [v.as_dict() for v in self.variantAttributes],
+            "variants": [v.as_dict() for v in self.variants],
+        }
```
Comment (IKarbowiak) [inline saleor/product/utils/variants_picker.py]: As above.

File: saleor/product/models.py [hunk 7]
```diff
@@ -148,7 +153,7 @@ def bulk_create(self, objs, batch_size=None, ignore_conflicts=False):
             objs, batch_size=batch_size, ignore_conflicts=ignore_conflicts
         )
 
-    def collection_sorted(self, user):
+    def collection_sorted(self, user: "User"):
```
Comment (pawelzar) [inline saleor/product/models.py]: Have you considered using `__future__.annotations`? https://www.python.org/dev/peps/pep-0563/
```suggestion
    def collection_sorted(self, user: User):
```

File: saleor/product/tasks.py [hunk 8]
```diff
@@ -33,35 +35,37 @@ def _update_variants_names(instance, saved_attributes):
...
+def update_variants_names(product_type_pk: List[str], saved_attributes_ids: List[str]):
     instance = ProductType.objects.get(pk=product_type_pk)
     saved_attributes = Attribute.objects.filter(pk__in=saved_attributes_ids)
-    return _update_variants_names(instance, saved_attributes)
+    _update_variants_names(instance, saved_attributes)
 
 
 @app.task
-def update_product_minimal_variant_price_task(product_pk):
+def update_product_minimal_variant_price_task(product_pk: str):
```
Comment (korycins) [inline saleor/product/tasks.py]: `str` or `int`?

File: saleor/product/utils/__init__.py [hunk 9]
```diff
@@ -55,46 +69,46 @@ def products_for_homepage(user, homepage_collection):
...
     return products
 
 
-def get_variant_url_from_product(product, attributes):
+def get_variant_url_from_product(product: "Product", attributes) -> str:
     return "%s?%s" % (product.get_absolute_url(), urlencode(attributes))
 
 
-def get_variant_url(variant):
+def get_variant_url(variant: "ProductVariant"):
```
Comment (korycins) [inline saleor/product/utils/__init__.py]: `-> str`?

File: saleor/product/tasks.py (hunk 8 above)
Comment (krzysztofwolski) [reply saleor/product/tasks.py]: when getting pks I'm used to strings

File: saleor/product/models.py (hunk 7 above)
Comment (krzysztofwolski) [reply saleor/product/models.py]: Did not consider that. I want to keep mostly known style, if that will be new standard then removing all `"` will be done by few clever regexp I think.

File: saleor/product/tasks.py (hunk 8 above)
Comment (korycins) [reply saleor/product/tasks.py]:
```
In [5]: type(Product.objects.all().first().id)
Out[5]: int
```

File: saleor/product/utils/variants_picker.py (hunk 2 above)
Comment (krzysztofwolski) [reply saleor/product/utils/variants_picker.py]:
```
>>> @dataclass
... class Z:
...   field: int
...
>>> @dataclass
... class A:
...   a_field: Z
...
>>> asdict(A(Z(3)))
{'a_field': {'field': 3}}
```

I swear to god it does not worked like this last week. 
I'll remove it then...

File: saleor/product/tasks.py (hunk 8 above)
Comment (krzysztofwolski) [reply saleor/product/tasks.py]: Bummer.

Comment (maarcingebala) [reply saleor/product/tasks.py]: Yep, as @korycins noticed database IDs should be rather typed as `int`.

File: saleor/product/utils/availability.py [hunk 10]
```diff
@@ -12,35 +11,40 @@
...
-    ),
-)
+if TYPE_CHECKING:
+    from django.db.models.query import QuerySet
+
+    from ...extensions.manager import ExtensionsManager
+
+
+@dataclass
+class ProductAvailability:
```
Comment (maarcingebala) [inline saleor/product/utils/availability.py]: At last we're getting rid of that namedtuple

**Pull request merged**